text/stacks/0000-node-on-tiny.md

```diff
@@ -0,0 +1,53 @@
+# Add to Tiny Run Image to Enable Node Apps
+
+## Summary
+
+At the moment, Node.js applications are not compatible with the tiny builder. Therefor the `nodejs` buildpack does not support the tiny stack.
+
+The tiny build image is already capable of building the Node.js application.
+
+The tiny run image is missing `libstdc++6`and `libgcc1`.
+
+## Motivation
+
+To support running Node.js applications with the tiny builder.
+
+While not all Node.js applications could be supported by that, this could be a feasible approach:
+
+- `tiny`: Run Node.js applications **without** any native extenstions
+- `base`: Run Node.js applications **without** many native extensions
+- `full`: Run Node.js applications **with** many native extensions
+
+This should be desirable as some types of Node.js application would benefit of the reduced image size.
+
+```bash
+apps/node-tiny                        latest     7546c8d45780   41 years ago   130MB
+apps/node-base                        latest     022cb9cea147   41 years ago   202MB
+apps/node-full                        latest     1324267830ec   41 years ago   803MB
+```
+
+## Detailed Explanation
+
+Install the [`libstdc++6`](https://packages.ubuntu.com/bionic/libstdc++6) and [`libgcc1`](https://packages.ubuntu.com/bionic/libgcc1) packages for the Ubuntu based tiny run image. It looks like this will add around 1.8 MB.
+
+```bash
+run-node-tiny                        latest     43acd28c97af   41 years ago   19.2MB
+paketobuildpacks/run:tiny            latest     fa34fc0b3d7b   41 years ago   17.4MB
+```
+
+## Rationale and Alternatives
+
+1. Node is written in C++, so it needs `libstdc++6`and `libgcc1` at runtime.
+
+## Implementation
+
+Install the `libstdc++6` and `libgcc1` packages for the Ubuntu based tiny run image.
+
+## Prior Art
+
+None
+
+## Unresolved Questions and Bikeshedding
+
+- Is it worth to influence other users of the stack?
+- Currently the `npm start` buildpack requires a `bash`, so any Node.js application using a start script would fail.
```

Comment:
I want to say yes to this, but these are the two points that I'm stuck on.

1. Adding this won't unlock the ability to run on Tiny for all Node.js use cases, in fact just a fraction because there are limitations like no start scripts & no native extensions. This leaves me wondering if it's worth it since there is an increase we're imposing on other users of about 10% in size.

    Perhaps you could expand more on what use cases this would enable in the motivations section. What situations or use cases are impractical with the present base stack, beyond image size? What could be unlocked by running on Tiny? 

2. I see the work upstream with the buildpacks project on extensions, which once implemented, would be an easy way to solve this problem. Given the increase in size of the tiny image, would it make more sense to recommend Node.js users wait for that work to be available instead of modifying Tiny.

Reply:
> 1. like no start scripts

We are currently evaluating a potential pr to `npm-start` to not use `bash`. This might open the possibilities to use start scripts with the tiny stack. I agree that without start scripts, the fraction of nodejs apps able to use this stack would probably be very low.

Reply:
3.) Would it be better to just have the Node.js dist buildpack manage this. For example, if the stack is Tiny. Then install the required libraries with the Node.js distribution. This would enable all the same Node.js use cases. We took a similar tactic in the Rust buildpack. Instead of including the library, it just checks if the stack is Tiny & then compiles out the need for the additional library.

Reply:
@dmikusa-pivotal if the approach you're describing is possible for Nodejs, I think that'd be an attractive option.

Reply:
My guess is that it isn't going to be easy to dynamically compile nodejs dependencies based on stack, because the nodejs distribution is precompiled, whereas it appears from a quick investigation that the rust distribution is compiled as part of the rustup buildpack's build phase.

Given that Stacks maintainers are happy with adding these libraries to the tiny stack (because we are making progress on creating a separate 'static' stack with fewer libraries), I'm inclined to resolve this conversation.

Any objections to resolving this conversation?

Reply:
It's not a matter of dynamically compiling, but just basically vendoring the required libraries with the buildpack. So if Node.js needs libstdc++6 and libgcc1 then you grab those and include them for all the supported stacks with the buildpack binaries.

At any rate, I'm OK resolving this because of other changes that have occurred recently.

Reply:
I am wondering what the difference between the current packages in tiny and the proposed to be added ones is? I like the idea of buildpacks just bringing their dependencies with them. Like we could use the new static stack for everything if the go-, java-, node-, python- etc. buildpacks would bring all libs they need, wouldn't we?